### ISSUE TITLE
ResultIterator should be canceled when parent context is done

### ISSUE THREAD

**Issue by mark-rushakoff** [influxdata/flux#306]
In our implementations in platform to handle tasks, we have code that could be simplified to roughly this:

```go
func ExecuteQuery(ctx context.Context, req *query.Request) {
	var svc query.QueryService = getQueryService()
	it, err := svc.Query(ctx, req)
	if err != nil {
		return err
	}
	defer it.Release()

	for it.More() { // it.More may block
		res := it.Next()
		if err := doSomethingWith(res); err != nil {
			return err
		}
	}
	return nil
```

We want this function to return when the passed-in context is canceled, but `it.More` may be in a blocking call.  See: https://github.com/influxdata/flux/blob/14b3100e9ede054cb420c5ea4f0880e539c9986a/result_iterator.go#L51

The temporary fix that I would like to apply is:

```go
func ExecuteQuery(ctx context.Context, req *query.Request) {
	var svc query.QueryService = getQueryService()
	it, err := svc.Query(ctx, req)
	if err != nil {
		return err
	}
	defer it.Release()

	// NEW CODE: Handle context closed on separate goroutine by releasing iterator:
	go func() {
		<-ctx.Done() // In this code path we know the context is always eventually canceled.
		it.Release() // ResultIterator godoc specifies: "It is safe to call Release multiple times."
	}()
	// END NEW CODE

	for it.More() { // it.More may block
		res := it.Next()
		if err := doSomethingWith(res); err != nil {
			return err
		}
	}
	return nil
```

However, the ResultIterator godoc also specifies:

>  all of the methods are expected to be called within the same goroutine.

And for platform's http.QueryService, calling Release from two goroutines causes a data race. I would prefer to see that constraint relaxed specifically for the Release method, but perhaps that is out of scope with context-aware result iterators.

I think the simplest additional API to the ResultIterator interface would be a `WithContext(context.Context) ResultIterator` method, which returns a new ResultIterator that respects the passed-in context's Done method; if ctx.Done() is closed, then:

- subsequent calls to ResultIterator.More() will return false
- the Result returned by an in-progress call to ResultIterator.Next() is valid, but its TableIterator will not return any new Tables after the current Table.
- ResultIterator.Err() will return ctx.Err()

I would also be satisfied with a new `ContextAwareResultIterator` type with method signatures that accept Context:

```go
type ContextAwareResultIterator interface {
    More(context.Context) bool
    Next(context.Context) Result
    Release()
    Err() error
    Statistics() Statistics
}
```

But it is somewhat strange to accept a context in Next and then return a Result that doesn't accept a context.

**Comment by mark-rushakoff** [comment]
I thought I did, but it was a bad test setup with a fake query service. 

If this is actually behaving as intended - i.e. cancelling the context passed to QueryService.Query, unblocks ResultIterator.More - then I think we still a godoc update for flux.Query to explain its behavior as it relates to a parent context. Maybe the godoc for `control/controller.Query` needs to be updated too.

**Comment by jlapacik** [comment]
I have been testing this issue, and I can confirm that `ResultIterator.More()` does not become unblocked when the parent query context is cancelled.

**Comment by jsternberg** [comment]
I can confirm that this is an issue with the Controller. It seems I found it and forgot to record it back when I was working on that a couple of weeks ago.

The Controller never reads the context passed in during creation so it doesn't cancel the query, but it should.

**Comment by jlapacik** [comment]
`More()` does not stay blocked when the parent query context is cancelled. It is released immediately. See #317 .
